dlog/kimchi/src/prover.rs

```diff
@@ -417,77 +386,180 @@ where
-        t8 += &emul8;
-        drop(emul8);
-        // endoscaling scalar computation
-        let emulscalar8 = endomul_scalar::constraint(range::ENDOMUL_SCALAR.start).evaluations(&env);
-        t8 += &emulscalar8;
-        drop(emulscalar8);
-        // poseidon
-        let pos8 = poseidon::constraint().evaluations(&env);
-        t8 += &pos8;
-        drop(pos8);
-
-        let t4 = match index.cs.chacha8.as_ref() {
-            None => t4,
-            Some(_) => {
-                let mut t4 = t4;
-                let chacha = chacha::constraint(range::CHACHA.start).evaluations(&env);
+            // permutation
+            let mut perm_alphas = all_alphas.get_alphas(ConstraintType::Permutation);
+            let (perm, bnd) = index
+                .cs
+                .perm_quot(&lagrange, beta, gamma, &z, &mut perm_alphas)?;
+            let mut t8 = perm;
+
+            if cfg!(test) {
+                let (_, res) = t8
+                    .clone()
+                    .interpolate()
+                    .divide_by_vanishing_poly(index.cs.domain.d1)
+                    .unwrap();
+                assert!(res.is_zero());
+            }
+
+            // scalar multiplication
+            let mut mul_alphas = all_alphas.take_powers(ConstraintType::Gate(GateType::Vbmul));
+            let mul8 = varbasemul::constraint(&mut mul_alphas).evaluations(&env);
+            t8 += &mul8;
+
+            if cfg!(test) {
+                let (_, res) = mul8
+                    .clone()
+                    .interpolate()
+                    .divide_by_vanishing_poly(index.cs.domain.d1)
+                    .unwrap();
+                assert!(res.is_zero());
+            }
+
+            drop(mul8);
+
+            // endoscaling
+            let mut endomul_alphas =
+                all_alphas.take_powers(ConstraintType::Gate(GateType::Endomul));
+            let emul8 = endosclmul::constraint(&mut endomul_alphas).evaluations(&env);
+            t8 += &emul8;
+
+            if cfg!(test) {
+                let (_, res) = emul8
+                    .clone()
+                    .interpolate()
+                    .divide_by_vanishing_poly(index.cs.domain.d1)
+                    .unwrap();
+                assert!(res.is_zero());
+            }
+
+            drop(emul8);
+
+            // endoscaling scalar computation
+            let mut endoscalar_alphas =
+                all_alphas.take_powers(ConstraintType::Gate(GateType::EndomulScalar));
+            let emulscalar8 = endomul_scalar::constraint(&mut endoscalar_alphas).evaluations(&env);
+            t8 += &emulscalar8;
+
+            if cfg!(test) {
+                let (_, res) = emulscalar8
+                    .clone()
+                    .interpolate()
+                    .divide_by_vanishing_poly(index.cs.domain.d1)
+                    .unwrap();
+                assert!(res.is_zero());
+            }
```

Comment:
added these tests as I was dealing with another issue that was hard to debug. These tests were such a time saver that I strongly think we should keep them in.

Reply:
sgtm!